File: day3/1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func getTreeMap() []string {
	file, _ := os.Open("./in")
	defer file.Close()
	scanner := bufio.NewScanner(file)

	var treeMap []string
	for scanner.Scan() {
		treeMap = append(treeMap, scanner.Text())
	}
	return treeMap
}

type Slope struct {
	Right int
	Down  int
}

func traveseBySlope(treeMap []string, slope Slope) int {
	width := len(treeMap[0])
	height := len(treeMap)
	y := 0
	x := 0
	count := 0

	for y+slope.Down < height {
		x = (x + slope.Right) % width
		y += slope.Down
		if treeMap[y][x] == '#' {
			count += 1
		}
	}
	return count
}

func execPart1(treeMap []string) int {
	slope := Slope{3, 1}
	count := traveseBySlope(treeMap, slope)
	return count
}

func getSlopeList() [5]Slope {
	slopes := [...]Slope{Slope{1, 1}, Slope{3, 1}, Slope{5, 1}, Slope{7, 1}, Slope{1, 2}}
	return slopes
}

func execPart2(treeMap []string) int {
	slopes := getSlopeList()

	product := 1
	for _, slope := range slopes {
		product *= traveseBySlope(treeMap, slope)
	}
	return product
}

func main() {
	treeMap := getTreeMap()

	// part 1
	treeCount := execPart1(treeMap)
	fmt.Println(treeCount)

	// part 2
	product := execPart2(treeMap)
	fmt.Println(product)
}

File: src/days/day3/1.py
import functools 

def read_map():
    f = open('in', 'r')
    content = []
    for x in f:
        content.append(x.strip('\n'))
    return content 


def traverse_by_slope(slope):
    tree_map = read_map()
    right, down = slope
    x=0
    y=0
    count=0
    height=len(tree_map)
    width=len(tree_map[0])
    

    while y + down < height:
        x+=right;
        x = x % width
        y+=down
        if tree_map[y][x] == '#':
            count+=1

    return count


# part 1
print(traverse_by_slope([3, 1]))

# part 2
slopes = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]

tree_count = map(traverse_by_slope, slopes)
product = functools.reduce(lambda a, b : a * b, tree_count)
print(product)
